### PR TITLE
[WIP][SPIKE] Don't bother with conventions if Sagas and TimeoutManager are disabled

### PR DESCRIPTION
**Pull request by DavidBoike**
DO NOT MERGE!

I want to find out if this approach is a fix for ServiceControl running Raven 3. Plan would be to close this PR and reintroduce as a hotfix.

/cc: @johnsimons @Particular/ravendb-persistence-maintainers

### CODE REVIEW

**Comment by mauroservienti** [comment]
Nice one, I like it.

File: src/NServiceBus.RavenDB.Tests/Persistence/DocumentIds/DocumentIdConventionTestBase.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ protected void ApplyTestConventions(IDocumentStore store, ConventionType type)
             ApplyConventionsInternal(store, type, false);
 
             var sagaTypes = new[] { typeof(TestSagaData) };
-            var conventions = new DocumentIdConventions(store, sagaTypes, EndpointName);
+            var conventions = new DocumentIdConventions(store, sagaTypes, EndpointName, true, true);
```
Comment (colin-higgins) [inline src/NServiceBus.RavenDB.Tests/Persistence/DocumentIds/DocumentIdConventionTestBase.cs]: Stupid nitpick, but I really prefer named parameters when there are more than a couple and especially when they're the same type.

**Comment by DavidBoike** [comment]
@johnsimons Can you pull down the [pull request NuGet from TeamCity](https://builds.particular.net/repository/download/NServiceBusRavenDB_Build/164796:id/nugets/NServiceBus.RavenDB%203.0.10-pullrequest0235.nupkg) and see if that fixes up ServiceControl? Or how would you like to proceed?

File: src/NServiceBus.RavenDB.Tests/Persistence/DocumentIds/DocumentIdConventionTestBase.cs (hunk 1 above)
Comment (DavidBoike) [inline src/NServiceBus.RavenDB.Tests/Persistence/DocumentIds/DocumentIdConventionTestBase.cs]: Yeah I'll make that clear, assuming this works for John and gets to release.

**Comment by DavidBoike** [comment]
@johnsimons Actually I got a hotfix-3.0.10 build ready, so you could try that instead:

https://builds.particular.net/repository/download/NServiceBusRavenDB_Build/164812:id/nugets/NServiceBus.RavenDB%203.0.10-beta0001.nupkg

Not enough time to turn them into PRs yet, plus I want to get the build servers ready to execute LINQPad scripts so we can clear out the test databases when these silly things run.

**Comment by johnsimons** [comment]
I'll test this today and report back

**Comment by johnsimons** [comment]
@DavidBoike looking at the fix, this should work for SC, but I it seems the exception will still occur if eg we enable timeouts ?
So I think this is a nice improvement but does not address #229

**Comment by DavidBoike** [comment]
@johnsimons I thought you said SC did not use sagas or timeouts?

**Comment by johnsimons** [comment]
> I thought you said SC did not use sagas or timeouts?

That is right @DavidBoike, but what happens if we decide to start using it ? We then will hit #229, right ?

File: src/NServiceBus.RavenDB/Internal/DocumentStoreInitializer.cs [hunk 2]
```diff
@@ -38,7 +38,9 @@ internal IDocumentStore Init(ReadOnlySettings settings)
 
         void ApplyConventions(ReadOnlySettings settings)
         {
-            var idConventions = new DocumentIdConventions(docStore, settings.GetAvailableTypes(), settings.EndpointName());
+            var sagasEnabled = settings.GetOrDefault<bool>(typeof(Features.Sagas).FullName);
```
Comment (johnsimons) [inline src/NServiceBus.RavenDB/Internal/DocumentStoreInitializer.cs]: @DavidBoike I know this works, but to me it feels like a hack that will bit us later on, IMO I would try to use a `Feature` with `Depends`.

**Comment by DavidBoike** [comment]
@johnsimons You're right, so the way I see it we have 3 options.

1. EmbeddedDocumentStore could never have been specified by connection string, so if the doc store is one, we forget the conventions completely and go with straight Raven defaults.
2. In 3.2 (already planned API update to make DTC configs easier) add a an advanced option `.NeverEverUseLegacyConventionsWhichIsOnlySafeForNewEndpoints()` and use that for ServiceControl.
3. Both 1 and 2.

Although in the freakish scenario that a customer tried to use an embedded DB on a normal endpoint, and then used one of the versions that created the doc of collections in use on a new database, so now has a very new embedded database configured for legacy conventions. If we are at all concerned about that, then option 2 is the only one on the table.

File: src/NServiceBus.RavenDB/Internal/DocumentStoreInitializer.cs (hunk 2 above)
Comment (DavidBoike) [inline src/NServiceBus.RavenDB/Internal/DocumentStoreInitializer.cs]: True. Can you make a feature that depends upon 2 other features being **disabled**?

Comment (johnsimons) [inline src/NServiceBus.RavenDB/Internal/DocumentStoreInitializer.cs]: I would actually make 2 features, one that depends on Sagas, and it adds the relevant mappings
and the other that depends on timeouts and adds the relevant mapping.

So this means that the "mappings dictionary" needs to be a singleton

**Comment by johnsimons** [comment]
@DavidBoike 
:+1: for option 2

**Comment by DavidBoike** [comment]
Cool, will run with it.

**Pull request closed without merging**